### PR TITLE
Create new command line engine

### PR DESCRIPTION
**Pull request by IntelOrca**
New command line engine and parser to replace argparse.c.

@LRFLEW Could you update xcode project?
@janisozaur @LRFLEW Could you test this with other platforms?
Thanks.

Suggestions welcome.

I should probably also note that installing scenarios or tracks is not yet implemented, even though it was available in vanilla.

### CODE REVIEW

File: src/cmdline/RootCommands.cpp [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+
+    CommandTableEnd
+};
+
+static const CommandLineExample CommandLine::RootExamples[]
+{
+    { "./my_park.sv6",                              "open a saved park"                      },
+    { "./SnowyPark.sc6",                            "install and open a scenario"            },
+    { "./ShuttleLoop.td6",                          "install a track"                        },
+    { "http:/openrct2.website/files/SnowyPark.sv6", "download and open a saved park"         },
```
Comment (janisozaur) [inline src/cmdline/RootCommands.cpp]: http:/`/`

**Comment by pizza2004** [comment]
Just as a question.  Would it be possible for @LRFLEW to PR his change onto this one and then you to just squash it into your commits, so that we don't have a commit just to add it to Xcode?  If so, that's possibly one way to handle things until we can possibly find a better way.

File: src/cmdline/RootCommands.cpp [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+{
+    { CMDLINE_TYPE_SWITCH,  &_help,            'h', "help",              "show this help message and exit"                            },
+    { CMDLINE_TYPE_SWITCH,  &_version,         'v', "version",           "show version information and exit"                          },
+    { CMDLINE_TYPE_SWITCH,  &_noInstall,       'n', "no-install",        "do not install scenario if passed"                          },
+    { CMDLINE_TYPE_SWITCH,  &_all,             'a', "all",               "show help for all commands"                                 },
+    { CMDLINE_TYPE_SWITCH,  &_about,           NAC, "about",             "show information about " OPENRCT2_NAME                      },
+    { CMDLINE_TYPE_SWITCH,  &_verbose,         NAC, "verbose",           "log verbose messages"                                       },
+    { CMDLINE_TYPE_SWITCH,  &_headless,        NAC, "headless",          "run " OPENRCT2_NAME " headless"                             },
+#ifndef DISABLE_NETWORK
+    { CMDLINE_TYPE_INTEGER, &_port,            NAC, "port",              "port to use for hosting or joining a server"                },
```
Comment (janisozaur) [inline src/cmdline/RootCommands.cpp]: This should mention "Use 11753 if in doubt" or some such.

**Comment by IntelOrca** [comment]
@pizza2004 yes that what he usually does.

**Comment by janisozaur** [comment]
@pizza2004, your suggestion is useless. this is done on a branch for a reason, see: https://github.com/OpenRCT2/OpenRCT2/pull/2680 and https://github.com/OpenRCT2/OpenRCT2/pull/2676

File: src/cmdline/RootCommands.cpp (hunk 2 above)
Comment (IntelOrca) [inline src/cmdline/RootCommands.cpp]: Well it defaults to config port which defaults to 11753

**Comment by pizza2004** [comment]
@IntelOrca I know he normally creates a pull request.  I carefully follow all progress on every ticket related to this project.  I was just wondering if it was possible for him to have his commit squashed into /this/ pull request as a way to avoid having a bunch of commits just to add single files to the Xcode project all the time.

@janisozaur What in the world are you even talking about?  Neither of those have anything at all to do with what I just said.  I'm honestly not even at all certain why you pointed them both out.  I was just making a suggestion on something to try to help keep the commit history a bit cleaner, so we don't have single commits just for the sake of the Xcode project whenever a new files is added.

**Comment by IntelOrca** [comment]
@pizza2004 Since when are single commits a problem? They are encouraged as they describe the code changes more concisely and make it easier to shift them around.

**Comment by pizza2004** [comment]
Oh definitely.  Any time the code base changes that makes sense, but the Xcode project just needs the file added to it.  Normally when you make a new file you add it to the Visual Studio project in the same commit yes?  I was just saying it would look cleaner to have it squashed into that same commit if possible.

File: src/cmdline/CommandLine.cpp [hunk 3]
```diff
@@ -0,0 +1,579 @@
...
+    }
+
+    static bool HandleSpecialArgument(const char * argument)
+    {
+#if defined(__APPLE__) && defined(__MACH__)
+        if (String::Equals(argument, "-NSDocumentRevisionsDebugMode"))
+        {
+            return true;
+        }
+        if (String::StartsWith("-psn_"))
```
Comment (pizza2004) [inline src/cmdline/CommandLine.cpp]: "Too few arguments to function call, expected at least 2, have 1" I get this error when trying to build.

Comment (LRFLEW) [inline src/cmdline/CommandLine.cpp]: I'm fixing that, don't worry. There's more errors than that, so I'm fixing all of them at once.

Comment (pizza2004) [inline src/cmdline/CommandLine.cpp]: I assumed you meant to put "argument," which worked.

File: src/cmdline/RootCommands.cpp [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+#endif
+
+    // Sub-commands
+    DefineSubCommand("screenshot", CommandLine::ScreenshotCommands),
+    DefineSubCommand("sprite",     CommandLine::SpriteCommands    ),
+
+    CommandTableEnd
+};
+
+static const CommandLineExample CommandLine::RootExamples[]
```
Comment (pizza2004) [inline src/cmdline/RootCommands.cpp]: "Static declaration of 'RootExamples' follows non-static declaration"  Xcode doesn't like this either.

File: src/cmdline/CommandLine.cpp (hunk 3 above)
Comment (pizza2004) [inline src/cmdline/CommandLine.cpp]: Oh, alright.  If you're already here I'll just leave it to you.

**Pull request merged**